### PR TITLE
Add base64 encoding and decoding APIs that accept az_span.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-c/issues/1618

cc @danewalton

### CODE REVIEW

File: sdk/src/azure/core/az_base64.c [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+// The maximum integer length of binary data that can be encoded into base 64 text and still fit
+// into an az_span which has an int32_t length, i.e. (INT32_MAX / 4) * 3;
+#define _az_MAX_SAFE_ENCODED_LENGTH 1610612733
+
+#define _az_ENCODING_PAD '='
+
+static const char _az_base64_encode_array[65]
+    = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+static const int8_t _az_base64_decode_array[256] = {
```
Comment (antkmsft) [inline sdk/src/azure/core/az_base64.c:18]: I would try save some of that 256 bytes and try to implement it with ifs/switch-cases.

File: sdk/src/azure/core/az_base64.c [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+#include <azure/core/az_base64.h>
+#include <azure/core/internal/az_precondition_internal.h>
+
+#include <azure/core/_az_cfg.h>
+
+// The maximum integer length of binary data that can be encoded into base 64 text and still fit
+// into an az_span which has an int32_t length, i.e. (INT32_MAX / 4) * 3;
+#define _az_MAX_SAFE_ENCODED_LENGTH 1610612733
+
+#define _az_ENCODING_PAD '='
```
Comment (antkmsft) [inline sdk/src/azure/core/az_base64.c:13]: Replace `#define` with unnamed enum?

File: sdk/src/azure/core/az_base64.c [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+
+static const char _az_base64_encode_array[65]
```
Comment (antkmsft) [inline sdk/src/azure/core/az_base64.c:15]: I think we use right-side const?

**Review by antkmsft — APPROVED** [review]

File: sdk/src/azure/core/az_base64.c [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+  -1,
+  -1,
+  -1,
+  -1,
+  -1,
+  -1,
+  -1,
+};
+
+static AZ_NODISCARD int32_t _az_base64_encode(uint8_t* three_bytes)
```
Comment (antkmsft) [inline sdk/src/azure/core/az_base64.c:278]: Mark a bunch of these statics as AZ_INLINE, especially if they are being used once?

Comment (ahsonkhan) [reply sdk/src/azure/core/az_base64.c:278]: That's an implementation detail/compiler optimization, I can do that separately.

File: sdk/src/azure/core/az_base64.c (hunk 1 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_base64.c:18]: That's an implementation detail, and not sure if it is worth it from a total code size perspective. I will investigate this in a follow-up:
https://github.com/Azure/azure-sdk-for-c/issues/1817

File: sdk/src/azure/core/az_base64.c (hunk 2 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_base64.c:13]: I saw this pattern being used in IoT already, so decided to follow it. We can change it across the board later if there is a coding guideline :)

**Review by danewalton — APPROVED** [review]

File: sdk/inc/azure/core/az_base64.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/inc/azure/core/az_base64.h:5]: I believe we got rid of that in most places already (probably the remaining IoT files can remove it too). Let me dig up the rational as to why.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/inc/azure/core/az_base64.h:5]: Here: https://github.com/Azure/azure-sdk-for-c/pull/1057

From @antkmsft 
> * Remove file name from the `@file` section - Doxygen still adds it to documentation perfectly, only that now if you rename a file, you don't need to update the comment, the name will get picked up automatically.

**Comment by ahsonkhan** [comment]
Code coverage looks good (almost 100% line coverage):
https://dev.azure.com/azure-sdk/public/_build/results?buildId=990064&_a=summary&tab=ms.vss-codecoverage-web.code-coverage-details&view=codecoverage-tab
image

The only branches missing is precondition testing:
https://github.com/Azure/azure-sdk-for-c/issues/1836

**Pull request merged**